### PR TITLE
Implement Kubernetes Client Code Generation & Increase Namespace Limit

### PR DESCRIPTION
**Pull request by andrewlecuyer**
This change formally integrates Kubernetes client code generation into the PostgreSQL Operator project.  Specifically, the `code-generator` project (https://github.com/kubernetes/code-generator) is now utilized to generate clients, informers and listers for use by the Operator containers, controllers, etc.

With the introduction of generated code, the various constructs created for creating clients and informers have now been leveraged throughout the Operator project.  For instance, 
 a `SharedInformerFactory` is now used to create and run the various informers for each controller, instead of creating them manually.  This approach ensures the Operator project aligns with and leverages the strategies for building a controller as demonstrated in the Kubernetes `sample-contoller` project (https://github.com/kubernetes/sample-controller).

Additionally, a new mechanism for creating and managing the various controllers created across all namespaces watched/managed by the Operator has now been implemented.  This provides a consistent way for tracking, starting and stopping various controllers throughout the lifetime of the Operator, e.g. adding/removing controllers as namespaces are added/removed.

Additionally, the controllers created within a single namespace now all utilize a single client, instead of leveraging a single client that is shared throughout each controller created by the Operator.  This will mitigate issues seen in environments in which a large number of namespaces are watched, e.g. throttling/rate-limiting that can occur when a single client connection exceeds a quota set by a specific cloud provider, delaying PG cluster creation and other administrative capabilities.  Specifically, this allows the Operator to watch and manage a much larger volume of namespaces without running into any limits, therefore mitigating the potential for delays when running various Operator logic.

And finally, also included are make targets with associated scripts for checking to see if any generated code needs to be updated, and then updating the generated code as needed.

**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**

- Informers are manually created
- A single client is shared among all controllers across all namespaces
- Controllers are not properly cleaned up when a namespace is removed

#1300
[ch7484]

**What is the new behavior (if this is a feature change)?**

- Code generation is used to generate client and informer code, which is now utilized for creating clients and informers
- Each namespace has a client for use by the controllers created for that namespace
- Controllers are properly cleaned up when a namespace is removed

**Other information**:
N/A

### CODE REVIEW

**Comment by andrewlecuyer** [comment]
Thanks for the feedback.

A `Code Generation` section has now been added to the developer docs.

File: controller/namespace/namespacecontroller.go [hunk 1]
```diff
@@ -16,139 +16,114 @@ limitations under the License.
...
+// NewNamespaceController creates a new namespace controller that will watch for namespace events
+// as responds accordingly.  This adding and removing controller groups as namespaces watched by the
+// PostgreSQL Operator are added and deleted.
+func NewNamespaceController(clients *kubeapi.ControllerClients,
+	controllerManager controller.ManagerInterface,
+	informer coreinformers.NamespaceInformer) (*Controller, error) {
+
+	controller := &Controller{
+		NamespaceClient:    clients.PGORestclient,
+		NamespaceClientset: clients.Kubeclientset,
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply controller/namespace/namespacecontroller.go]: Really I was just trying to create some consistency around, our various controllers (e.g. each obtains both a Kube & PGO client), but at this point the namespace controller really no longer needs any client itself, so they can be removed.

File: kubeapi/client_config.go [hunk 2]
```diff
@@ -33,24 +43,65 @@ func loadClientConfig() (*rest.Config, error) {
...
+	return kubernetes.NewForConfig(config)
+}
+
+// createPGOClient creates a PGO RESTClient and Clientset using the provided configuration
+func createPGOClient(config *rest.Config) (*rest.RESTClient, *clientset.Clientset, error) {
+	// create a client for pgo resources
+	pgoClientset, err := clientset.NewForConfig(config)
+	pgoRESTClient := pgoClientset.CrunchydataV1().RESTClient().(*rest.RESTClient)
+	if err != nil {
+		log.Error(err)
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply kubeapi/client_config.go]: Log has been removed.

File: pgo-scheduler/pgo-scheduler.go [hunk 3]
```diff
@@ -172,86 +181,35 @@ func newKubeClient() (*kubernetes.Clientset, error) {
 	return client, nil
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply pgo-scheduler/pgo-scheduler.go]: `newKubeClient` has been removed.

File: pgo-rmdata/pgo-rmdata.go [hunk 4]
```diff
@@ -61,12 +60,12 @@ func main() {
 	}
 
 	var err error
-	request.RESTConfig, request.Clientset, err = kubeapi.NewClient()
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply pgo-rmdata/pgo-rmdata.go]: Good catch...`RESTConfig` is now once again assigned.

File: pgo-rmdata/pgo-rmdata.go [hunk 5]
```diff
@@ -61,12 +60,12 @@ func main() {
...
+	_, request.Clientset, err = kubeapi.NewKubeClient()
 	if err != nil {
 		log.Fatalln(err.Error())
 	}
 
-	request.RESTClient, _, err = util.NewClient(request.RESTConfig)
+	_, request.RESTClient, _, err = kubeapi.NewPGOClient()
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply pgo-rmdata/pgo-rmdata.go]: Yup, looks like now that we're grabbing the config per your feedback above, the call to `NewKubeClient` and `NewPGOClient` should now populate both clients, along with a copy of the config.

File: pgo-scheduler/scheduler/configmapcontroller.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	if err := c.Scheduler.AddSchedule(cm); err != nil {
+		log.WithFields(log.Fields{
+			"error": err,
+		}).Error("Failed to add schedules")
+	}
+}
+
+// onDelete is called when a configMap is deleted
+func (c *Controller) onDelete(obj interface{}) {
+	{
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply pgo-scheduler/scheduler/configmapcontroller.go]: Extra brackets removed.

File: hack/boilerplate.go.txt [hunk 7]
```diff
@@ -0,0 +1,15 @@
+/*
+ Copyright 2020 Crunchy Data Solutions, Inc.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
```
[the remark this answers is not in the document]
Comment (jkatz) [reply hack/boilerplate.go.txt]: The convention has been to not have the lead space, it appears. I'm not...overly concerned about the spacing. The format of the copyright is the one I have opinions on given the number of hours it took for me to standardize it

File: postgres-operator.go [hunk 8]
```diff
@@ -57,117 +49,59 @@ func main() {
 	//give time for pgo-event to start up
 	time.Sleep(time.Duration(5) * time.Second)
 
-	config, Clientset, err := kubeapi.NewControllerClient()
+	clients, err := kubeapi.NewControllerClients()
 	if err != nil {
 		log.Error(err)
 		os.Exit(2)
 	}
+	log.Debugf("operator clients created")
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply postgres-operator.go]: Guess we don't really need that, I'm good w/removing.

File: postgres-operator.go [hunk 9]
```diff
@@ -57,117 +49,59 @@ func main() {
...
-	jobcontroller := &job.Controller{
-		JobConfig:          config,
-		JobClientset:       Clientset,
-		JobClient:          crdClient,
-		Ctx:                ctx,
-		InformerNamespaces: make(map[string]struct{}),
+	// create a new controller manager with controllers for all current namespaces and then run
+	// all of those controllers
+	controllerManager, err := manager.NewControllerManager(namespaceList)
+	controllerManager.RunAll()
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply postgres-operator.go]: Error is now properly handled.

File: hack/update-codegen.sh [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+SCRIPT_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+CODEGEN_PKG=${CODEGEN_PKG:-$(cd "${SCRIPT_ROOT}"; ls -d -1 ./vendor/k8s.io/code-generator 2>/dev/null || echo ../code-generator)}
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply hack/update-codegen.sh]: I borrowed this from the sample controller, so was just trying to stick with that template:

https://github.com/kubernetes/sample-controller/blob/041853156b367ba768df5cda747e653dd98a7127/hack/update-codegen.sh#L22

File: pgo-scheduler/pgo-scheduler.go [hunk 11]
```diff
@@ -138,11 +137,21 @@ func main() {
...
+	controllerManager, err := sched.NewControllerManager(namespaceList, scheduler)
+	if err != nil {
+		log.WithFields(log.Fields{}).Fatalf("Failed to create controller manager: %s", err)
+		os.Exit(2)
 	}
+	controllerManager.RunAll()
 
-	SetupNamespaceWatch(installationName, scheduler, stop)
+	nsKubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, 0)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply pgo-scheduler/pgo-scheduler.go]: This is the informer factory utilized to create the namespace controller for the scheduler a few lines down.

File: controller/manager/controllermanager.go [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+	pgoInformerFactory := informers.NewSharedInformerFactoryWithOptions(pgoClientset, 0,
+		informers.WithNamespace(namespace))
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(kubeClientset, 0,
+		kubeinformers.WithNamespace(namespace))
+
+	pgTaskcontroller := &pgtask.Controller{
+		PgtaskConfig:    config,
+		PgtaskClient:    pgoRESTClient,
+		PgtaskClientset: kubeClientset,
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply controller/manager/controllermanager.go]: The current conventions for naming the vars in the various controller structs are a bit confusing right now (and I did not change them for this specific PR), but the `Clientset` in these structs is a Kubernetes `ClientSet`, and the `pgoRESTClient` is utilized for access PGO resources.  E.g. you'll see that in the current `pgtask` controller its a Kubernetes `clientset`:

https://github.com/CrunchyData/postgres-operator/blob/547180692763640a5141e3aacd2c2e09eacce3ac/controller/pgcluster/pgclustercontroller.go#L49

File: controller/manager/controllermanager.go [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+	pgPolicycontroller := &pgpolicy.Controller{
+		PgpolicyClient:    pgoRESTClient,
+		PgpolicyClientset: kubeClientset,
+		Informer:          pgoInformerFactory.Crunchydata().V1().Pgpolicies(),
+	}
+
+	podcontroller := &pod.Controller{
+		PodConfig:    config,
+		PodClientset: kubeClientset,
+		PodClient:    pgoRESTClient,
```
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply controller/manager/controllermanager.go]: The `PodClient` is utilized for accessing PGO resources, so the `pgoRESTClient` is applicable here.

File: controller/namespace/namespacecontroller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply controller/namespace/namespacecontroller.go]: Clients have been removed.

File: hack/update-codegen.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply hack/update-codegen.sh]: Script updated.

File: controller/manager/controllermanager.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply controller/manager/controllermanager.go]: Basically each controller (whether for a Kube resource or a PGO resource) has the potential need to access both PGO and Kube resources, so each gets a client for both: `kubeClientset` for kube resources & `pgoRESTClient` for PGO resources.

File: hack/boilerplate.go.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply hack/boilerplate.go.txt]: Good catch, I'll fix the spacing and re-generate to bring the generated code in line with all other go code.

[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply hack/boilerplate.go.txt]: ...although I'm seeing that we're not 100% consistent.  @jkatz I'll let you weigh in on the proper/desired formatting.

[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply hack/boilerplate.go.txt]: Sorry, realized I was looking at an old view and now see you already responded :)

I'll go ahead and remove the space.

[the remark this answers is not in the document]
Comment (andrewlecuyer) [reply hack/boilerplate.go.txt]: Headers have been updated.

File: pgo-scheduler/pgo-scheduler.go [hunk 14]
```diff
@@ -157,101 +165,35 @@ func main() {
...
-			AddFunc: func(obj interface{}) {
-				ns, ok := obj.(*v1.Namespace)
-				if !ok {
-					log.WithFields(log.Fields{}).Error("Could not convert runtime object to Namespace..")
+// SetupNamespaceController sets up a namespace controller that monitors for namespace add and
+// delete events, and then either creates or removes controllers for those namespaces
+func SetupNamespaceController(installationName string, scheduler *scheduler.Scheduler,
+	informer coreinformers.NamespaceInformer, controllerManager controller.ManagerInterface) {
+
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```
Comment (cbandy) [inline pgo-scheduler/pgo-scheduler.go:173]: Probably need an `UpdateFunc`, too.

File: controller/namespace/namespacecontroller.go [hunk 15]
```diff
@@ -16,139 +16,106 @@ limitations under the License.
...
 	}
 
+	log.Debugf("namespace Controller: onAdd crunchy namespace %s created", newNs.ObjectMeta.SelfLink)
+	c.ControllerManager.AddAndRunControllerGroup(newNs.Name)
 }
 
-// onUpdate is called when a pgcluster is updated
+// onUpdate is called when a namespace is updated
 func (c *Controller) onUpdate(oldObj, newObj interface{}) {
-	//oldNs := oldObj.(*v1.Namespace)
```
Comment (cbandy) [inline controller/namespace/namespacecontroller.go:122]: What about when `oldNs` was being watched but `newNs` wouldn't? Seems like we should `RemoveGroup()` in that case.

File: postgres-operator.go [hunk 16]
```diff
@@ -57,117 +49,62 @@ func main() {
...
-		JobConfig:          config,
-		JobClientset:       Clientset,
-		JobClient:          crdClient,
-		Ctx:                ctx,
-		InformerNamespaces: make(map[string]struct{}),
+	controllerManager.RunAll()
+	log.Debug("controller manager created and all included controllers are now running")
+
+	nsKubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClientset, 0)
+	if err != nil {
```
Comment (cbandy) [inline postgres-operator.go:87]: This is the same `err` from L78-79. What should be checked after calling `kubeinformers.NewSharedInformerFactory` ?

File: pgo-scheduler/pgo-scheduler.go [hunk 17]
```diff
@@ -138,11 +136,21 @@ func main() {
...
+	controllerManager, err := sched.NewControllerManager(namespaceList, scheduler)
+	if err != nil {
+		log.WithFields(log.Fields{}).Fatalf("Failed to create controller manager: %s", err)
+		os.Exit(2)
 	}
+	controllerManager.RunAll()
 
-	SetupNamespaceWatch(installationName, scheduler, stop)
+	nsKubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, 0)
+	if err != nil {
```
Comment (cbandy) [inline pgo-scheduler/pgo-scheduler.go:147]: This `err` was already checked on line L140. What should be checked after calling `kubeinformers.NewSharedInformerFactory` ?

**Pull request merged**